### DISCUSSION TITLE
What about running echobot python script as a service ?

### DISCUSSION THREAD

**Discussion by vincenzofrezza** [python-telegram-bot/python-telegram-bot#2498] [Q&A]
Is it possibile to run the echobot.py, included in examples of python-telegram-bot, as a linux service on an ubuntu instance ?

I tried to do it but without success.

I followed indication [found here](https://websofttechs.com/tutorials/how-to-setup-python-script-autorun-in-ubuntu-18-04/)  but when i start the service it fails to run.

This is the result of _sudo systemctl status tlgrm-echobot.service_

```
tlgrm-echobot.service - Telegram echo bot service
     Loaded: loaded (/lib/systemd/system/tlgrm-echobot.service; enabled; vendor preset: enabled)
     Active: failed (Result: exit-code) since Thu 2021-04-29 17:45:35 CEST; 1min 3s ago
    Process: 180656 ExecStart=/usr/bin/python /home/ubuntu/python/python-telegram-bot/echobot.py (code=exited, status=203/EXEC)
   Main PID: 180656 (code=exited, status=203/EXEC)
```

Here is  the content of _/usr/bin/tlgrm_echobot.service_

[Unit]
Description=Telegram echo bot service
After=multi-user.target
Conflicts=getty@tty1.service

[Service]
Type=simple
ExecStart=/usr/bin/python /home/ubuntu/python/python-telegram-bot/echobot.py
StandardInput=tty-force

[Install]
WantedBy=multi-user.target

Anyway, may target is to find the best way to run always a python script that manages my new telegram bot I created, not so far from echobot.py script.

Thanks in advance fon any kind of help, Vincenzo Frezza.

**Comment by Poolitzer** [comment]
use `journalctl -xe` to see the error why it didnt run.

**Reply by vincenzofrezza** [reply]
Thanks @Poolitzer for your hint but the result of journalctl -xe is really poor for me:

`-- The unit tlgrm-echobot.service has entered the 'failed' state with result 'exit-code'.
May 04 10:47:54 ip-172-31-16-15 sudo[268255]:   ubuntu : TTY=pts/0 ; PWD=/home/ubuntu ; USER=root ; COMMAND=/usr/bin/systemctl status tlgrm-echobot.service
May 04 10:47:54 ip-172-31-16-15 sudo[268255]: pam_unix(sudo:session): session opened for user root by ubuntu(uid=0)
May 04 10:47:54 ip-172-31-16-15 sudo[268255]: pam_unix(sudo:session): session closed for user root
May 04 10:48:01 ip-172-31-16-15 CRON[268260]: pam_unix(cron:session): session opened for user ubuntu by (uid=0)
May 04 10:48:01 ip-172-31-16-15 CRON[268259]: pam_unix(cron:session): session opened for user ubuntu by (uid=0)
`

No indications to investigate more from this result
I am looking for another way to accomplish my task, for example using Upstart instead of service daemon...

Thanks, Vincenzo

**Reply by Poolitzer** [reply]
hm thats weird, maybe the logging gets swallowed somewhere. try running
`/usr/bin/python /home/ubuntu/python/python-telegram-bot/echobot.py` from the ssh connection

**Reply by Poolitzer** [reply]
IB4 plot twist this links to the python2 installation on the system

**Reply by vincenzofrezza** [reply]
Rigth suggestions, this morning I changed the ExceStart in
`ExecStart=/usr/bin/python3 /home/ubuntu/python/python-telegram-bot/echobot.py`

Now the tlgrm-echobot.service is

`[Unit]
Description=Telegram echo bot service
After=multi-user.target
Conflicts=getty@tty1.service

[Service]
Type=simple
ExecStart=/usr/bin/python3 /home/ubuntu/python/python-telegram-bot/echobot.py
StandardInput=tty-force

[Install]
WantedBy=multi-user.target
`
After
`sudo systemctl daemon-reload`
`sudo systemctl enable tlgrm-echobot.service`
`sudo systemctl status tlgrm-echobot.service`

...the service fails :-(

`
● tlgrm-echobot.service - Telegram echo bot service
     Loaded: loaded (/lib/systemd/system/tlgrm-echobot.service; enabled; vendor preset: enabled)
     Active: failed (Result: exit-code) since Tue 2021-05-04 16:19:39 CEST; 1min 19s ago
   Main PID: 272130 (code=exited, status=1/FAILURE)

May 04 16:19:39 ip-172-31-16-15 systemd[1]: Started Telegram echo bot service.
May 04 16:19:39 ip-172-31-16-15 systemd[1]: tlgrm-echobot.service: Main process exited, code=exited, status=1/FAILURE
May 04 16:19:39 ip-172-31-16-15 systemd[1]: tlgrm-echobot.service: Failed with result 'exit-code'.
`
Vincenzo

**Reply by Poolitzer** [reply]
Hm, I would start writing logs to a log file, and see what they actually say when you start the file via a service

**Comment by vincenzofrezza** [comment]
I found a solution, a quick and dirty one ...

Simply I run the command using nohup

`nohup python3 echobot.py &`

Not the best solution but I have got time to investivate further ...

**Comment by s-raza** [comment]
Here is what works for a Bot I run in a production environment. This works without any flaws on the live server.

I am using pipenv so I have the full path to the Python interpreter in  ExecStart and WorkingDirectory pointed to the directory that has the main entry point to the Telegram bot.

```
[Unit]
Description=TelegramBot Service
After=multi-user.target

[Service]
Type=simple
ExecStart=/python/.venvs/vendordb-iWxSLQwW/bin/python ./start_bot.py -l d -ld '/logs/telegram/'
User=botservice
WorkingDirectory=/system/telegrambot
Restart=on-failure

[Install]
WantedBy=multi-user.target
Alias=telegrambot
```
